src/merkle/mmr/partial.rs

```diff
@@ -123,14 +125,37 @@ impl PartialMmr {
         }
     }
 
-    // MODIFIERS
+    // ITERATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// TODO:
+    pub fn inner_nodes(&self) -> impl Iterator<Item = InnerNodeInfo> + '_ {
+        let mut inner_nodes = Vec::new();
+        for (index, &value) in self.nodes.iter() {
+            if index.is_leaf() {
+                let mut idx = *index;
+                let mut nodes = Vec::new();
+                while let Some(node) = self.nodes.get(&idx.sibling()) {
+                    nodes.push(*node);
+                    idx = idx.parent();
+                }
+                let path = MerklePath::new(nodes);
+                let index = index.inner() / 2;
+                inner_nodes.extend(path.inner_nodes(index, value).unwrap());
+            }
+        }
+
+        inner_nodes.into_iter()
```

Comment:
This is currently a very inefficient implementation (and will produce duplicate results). Will refactor this tomorrow.

Reply:
I don't know what is the use case of this, so I don't know what you trying to optimize for, here are a few notes:

- If this `inner_nodes` is not called often, it may be okay to pay the cost of recomputing the inner nodes (this is by definition sparse, so it shouldn't be a lot of hashing anyways). In this case, here is one proposal:
  - The `inner_nodes` method receives a list with `(value, idx)`.
  - The list must be ordered by **decreasing** `idx`, this is a working stack.
  - loop while the list is not empty:
    - pop the last `(value, idx)`, compute `parent`, and yield the inner node
    - here i can think of two options:
        - have another vector, and push the newly computed `parent` as todo work, this will traverse the tree in level order
        - use a preorder-index and push it to the vector, this will traverse the tree in reverse preorder
        - both of the pushes above should preserve ordering and check for duplicates using bissect
- if the function is called very often, and we want to amortize the cost of hashing, than IMO a new structure should be created

Reply:
I described the use case briefly in https://github.com/0xPolygonMiden/crypto/pull/238#discussion_r1431742754. Overall, I think the first approach should be fine - i.e., re-computing inner nodes - as I don't expect it to be used very heavily (i.e., we'll need to use it to prove a transaction, but in this context, calling `inner_nodes()` is negligible).